### PR TITLE
Decoupling: microVM lifetime clearly split between pre-boot and post-boot.

### PR DESCRIPTION
**Pull request by acatangiu**
## Description of Changes

Disclaimer: This is a rather heavy PR/RFC, but benefits should be obvious.

The microVM lifetime and how resources are owned and managed is now clearly split in two:
1. There is a pre-boot *vmm-building* loop where pre-boot-supported API requests are handled. These mostly work on the resource collection defining the microVM. At this time, Firecracker acts as a data-store exposing a REST API for managing the data (microVM resources). A successful `StartMicroVm` command (successful boot) will move the microVM into the post-boot state, while failures are recoverable as they do not dirty/mutate the data.
2. There is a post-boot `VmController` which runs a control loop that handles `Vmm` internal events and also handles post-boot-supported API requests.

There is a very clear separation of concerns between objects used in each of these consecutive stages.

Another clear delimitation which occurred is between:
1. Configuring from API
and
2. Configuring from cmdline JSON

Another one is between a microVM with its API enabled or disabled.

The `VmController` was further enhanced to provide a generic interface taking an external `trait` object for externally handling control events.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/firecracker/src/api_server_adapter.rs [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                        seccomp_level,
+                        epoll_context,
+                    );
+
+                    // Send back the result.
+                    self.to_api
+                        .send(response)
+                        .map_err(|_| ())
+                        .expect("one-shot channel closed");
+                    built_vmm = maybe_vmm;
```
Comment (sandreim) [inline src/firecracker/src/api_server_adapter.rs]: I do not personally like maybe_vmm. I would prefer that handle_preboot_requests would return a response type which contains an enum. Matching that enum to vmm built result should provide the loop exit signal.

**Comment by alexandruag** [comment]
Reviewing, please don't merge yet.

File: src/firecracker/src/api_server_adapter.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+        to_api: Sender<VmmResponse>,
+    ) -> Self {
+        ApiAdapter {
+            api_event_fd,
+            from_api,
+            to_api,
+        }
+    }
+
+    /// Builds and starts a microVM. It either uses a config json or receives API requests to
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/firecracker/src/api_server_adapter.rs]: As far as I can see, this function returns when the microVM is built and started (it needs a `StartMicroVm` action in order to break out from the loop).

File: src/vmm/src/resources.rs [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+use vmm_config::net::*;
+use vmm_config::vsock::*;
+
+type Result<E> = std::result::Result<(), E>;
+
+/// Errors encountered when configuring microVM resources.
+#[derive(Debug)]
+pub enum Error {
+    /// JSON is invalid.
+    InvalidJson,
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/vmm/src/resources.rs]: How it is different from `InvalidJson` error?

[the remark this answers is not in the document]
Comment (sandreim) [reply src/vmm/src/resources.rs]: I was asking if the JSON parsing produces meaningful error messages that we can forward to the user so the error message has more value.

[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/vmm/src/resources.rs]: I agree. It would be helpful if we can provide meaningful parse error messages to users.

**Comment by sandreim** [comment]
> I think there's one high level aspect that should be addressed at this point, namely the Vmm becoming Firecracker-specific again. One of the nicer outcomes of decoupling is having a Vmm agnostic to Firecracker devices, the way we setup memory, etc, which can be used for different use cases based on a generic interface. One example approach is the current state of using `MmioDevice` instead of specific devices when building the Vmm, and something like the [VmmBuilderConfig](https://github.com/firecracker-microvm/firecracker/blob/43d2341e842b85cd77f23c18d2fa087d03ebbf9b/src/vmm/src/lib.rs#L399) (naming aside) struct for the initial configuration.

Further incremental PR's should completly decouple the VMM from it's devices. One way of doing that would be to implement a new DeviceManager which would be responsible to own and operate the devices. The VMM would only need to hold a mutable ref to a DeviceManager trait object. As we go further we will need to identify options for building and registering the devices without needing the VMM instance like we do now.

> 
> The newly proposed builder looks tied to a `VmResources` object, which in turn relies on Firecracker specific configuration (I assumed the builder should belong to the generic part of the Vmm logic, since there's no other interface for creating a Vmm). Moreover, I think the `GuestMemory` object passed to the Vmm should be passed to the generic logic from the outside (including loading the kernel image), since there are multiple ways of getting a fully initialized `GuestMemory` object besides the way it happens in Firecracker right now. In particular, when we start using `vm-memory`, the Vmm should be able to work with any guest memory object that implements the `GuestMemory` trait.

This is a fair point and I believe we can postpone this up until later when we will look at merging code back to the master branch.

> 
> Otherwise, I think we can move ahead with these changes and postpone discussing other details until we get a better idea about the polly integration (which will probably require additional changes anyway) and we start bringing things into the master branch.

**Comment by acatangiu** [comment]
I've synced up offline with @alexandruag and agreed this PR takes things in the right direction and through iteration we can address any outstanding issues.

**Pull request merged**